### PR TITLE
OCPBUGS-20097: Migrate tools image to RHEL9

### PR DESCRIPTION
**Pull request by soltysh**
/assign @ardaguclu

### CODE REVIEW

**Comment by ardaguclu** [comment]
/hold
Failures are related and no point of retesting

**Comment by ardaguclu** [comment]
`gssapi/gssapi.h: No such file or directory` looks like a blocker. I wonder what should we do for such errors.

**Comment by soltysh** [comment]
> `gssapi/gssapi.h: No such file or directory` looks like a blocker. I wonder what should we do for such errors.

we need to fix our builder, just looking where we're missing `krb5-devel` rpm :wink:

**Comment by ardaguclu** [comment]
> > `gssapi/gssapi.h: No such file or directory` looks like a blocker. I wonder what should we do for such errors.
> 
> we need to fix our builder, just looking where we're missing `krb5-devel` rpm 😉

Good to know that this is something we can fix our own. Thanks for the answer.

**Comment by soltysh** [comment]
So far I've struggled with the following issues:

**Comment by ardaguclu** [comment]
To fix ipv6 failure, we need to extract `oc.rhel8` in 4.15 and 4.16 clusters https://github.com/openshift-metal3/dev-scripts/blob/d7d7a3b867fe4a6446d4cb5430a99cc1f105e382/ocp_install_env.sh#L39

**Comment by soltysh** [comment]
> To fix ipv6 failure, we need to extract `oc.rhel8` in 4.15 and 4.16 clusters https://github.com/openshift-metal3/dev-scripts/blob/d7d7a3b867fe4a6446d4cb5430a99cc1f105e382/ocp_install_env.sh#L39

Fix in https://github.com/openshift-metal3/dev-scripts/pull/1646

**Comment by ardaguclu** [comment]
seems unrelated;
/test e2e-metal-ipi-ovn-ipv6

File: images/recycler/Dockerfile.rhel [hunk 1]
```diff
@@ -1,4 +1,4 @@
-FROM registry.ci.openshift.org/ocp/4.11:cli
+FROM registry.ci.openshift.org/ocp/4.16:cli
```
Comment (joepvd) [inline images/recycler/Dockerfile.rhel:1]: This dockerfile is not used in production builds

File: images/tools/Dockerfile [hunk 2]
```diff
@@ -1,4 +1,10 @@
-FROM registry.ci.openshift.org/ocp/4.16:cli
+FROM registry.ci.openshift.org/ocp/builder:rhel-9-golang-1.21-openshift-4.16 AS builder
+WORKDIR /go/src/github.com/openshift/oc
+COPY . .
+RUN make build --warn-undefined-variables
+
+COPY --from=builder /go/src/github.com/openshift/oc/oc /usr/bin/
```
Comment (joepvd) [inline images/tools/Dockerfile:6]: There is a FROM missing here

Comment (soltysh) [reply images/tools/Dockerfile:7]: Good catch - fixed.

File: images/recycler/Dockerfile.rhel (hunk 1 above)
Comment (soltysh) [reply images/recycler/Dockerfile.rhel:1]: Yeah, I bet some should be cleaned up

**Comment by ardaguclu** [comment]
https://prow.ci.openshift.org/view/gs/test-platform-results/pr-logs/pull/openshift_oc/1652/pull-ci-openshift-oc-master-e2e-aws-ovn/1775511373447434240 failed with `/test/curl: /lib64/libc.so.6: version `GLIBC_2.33' not found (required by /test/curl)`

**Comment by ardaguclu** [comment]
unrelated;
/test e2e-metal-ipi-ovn-ipv6

**Comment by ardaguclu** [comment]
reminder: Bot reverted https://github.com/openshift/release/pull/50563, we need to reintroduce our changes before retesting.

**Comment by ardaguclu** [comment]
Opened this https://github.com/openshift/release/pull/50581

**Comment by ardaguclu** [comment]
There needs to be a rhel9 change in oc in release repo which was reverted by bot.

**Comment by soltysh** [comment]
/hold
to ensure we don't merge before green tests

**Comment by soltysh** [comment]
This proofed to work
/hold cancel
/test e2e-aws-ovn-serial

**Comment by soltysh** [comment]
/override ci/prow/e2e-aws-ovn-serial
this failed b/c the changes from https://github.com/openshift/release/pull/51176 got already reverted by the bot

**Pull request merged**